### DISCUSSION TITLE
Feature Freeze for Release 6.0 and deferred refactor work?

### DISCUSSION THREAD

**Discussion by rocky** [Mathics3/mathics-core#660] [General]
It has been a while since the last release.

Also, there has been a temporary lull in activity and it is holiday season.

I would like to propose a refactor freeze and focus on just some cleanup, and then a release of the existing code. 

And then a continuance of some of the deeper issues that plague our code. 

Here are some of the kinds of cleanup, that I don't think are controversial.

* eval methods rename to `eval`. 
* eval code moved (to `mathics.eval`) from Builtin class definitions (in `mathics.core.builtin`) 
* pytest code inside builtin docstrings, (these are start with `#>` into pytest.
* builtin top-level pytest code in `pytest` moved to `pytest/builtin`
* Greater and more consistent use of URL to both WMA links and Wikipedia links
* Reduction of strings when Symbols are intended and appropriate
* More type annotations on functions and methods
* More docstrings on functions and methods
* [isort](https://pycqa.github.io/isort/)  imports, 

Here of some of the refactor things currently underway, that I proposed that we defer work on until after 6.0

* Boxing refactoring
    - MakeBox refactoring
    - Specialization of MakeBox `evaluate()`
    - Boxing rule correction
* RowBox, SuperscriptBox, SubscriptBox, UnderscriptBox, FractionBox, RadicalBox, FormBox  parsing shortcuts `\(`, `\+, `\&`,  `\%` `\@` - these also interact with Boxing refactoring
* Further performance refactoring
   - Addition of List and Association literal types
   - Overhaul of rule and Pattern matching
* Further Assignment cleanup (should possibly be based on LHS rather than on specific builtin name)
* Moving code out of `builtin/__init__.py`  `Definitions.__init__`  in preparation for being able to write and load an interpreter image. (This is under the category of speeding up loading). 

And there is longer term stuff that I had always wanted to work on like:
*  Turning the tree evaluation into more of an instruction interpreter (which would allow JITing and much better compilation).
*  Being able to translate Mathics into pure SymPy Python

As for the method. As is usual. The first part would be to update the docker images and let them sit for a week. I hope to have this done sometime around the New Year.  With that in place we could do Pip releases of all of the Python packages. 

Thoughts?

**Comment by TiagoCavalcante** [comment]
@rocky I agree in all topics (but again I don't think I'll can help with code :disappointed:, only PR reviews)

**Comment by rocky** [comment]
A little status... at one time I had thought we could have a release by the New Year. Now I am not certain. 

Up until recently we had been focusing on some of the larger structural problems with the code which is on hold until after release. 

However I realize that there are numerous cleanup kinds of things also. And were these done, it would make refactoring easier. I forgot about how the size of what could easily be cleaned up and slightly improved.

**Comment by rocky** [comment]
Status: We are closer to being able to release Mathics3 core. The most code as of Sunday is available on dockerhub.

However, I see that there was some cleanup work with respect to PyMathics modules that was deferred from the last release, so I am currently thinking of spending a couple of weeks cleaning that up.

**Comment by rocky** [comment]
Release 6.0.0 for everything except mathics-omnibus is on PyPI and as a github release. 

Thanks to Mauricio and Tiago for the long and hard work on this. 

Some specific things. 

mathics-core 6.0.0 mistakenly omitted Pillow as requirement for Python > 3.6.  So there may be a minor 6.0.1 release to correct this in a couple of weeks. If there are other small bug fixes that can go in too. 

For mathics-omnibus, release-6.0.0 is the branch that when merged will be the 6.0.0 release. I think the Python install portion of this is okay.  Building the docker image the main problem is that we are not finding spacey word libraries.  This is a relatively small item though. 

 When that gets cleared up I'll make a release of that.

**Comment by rocky** [comment]
There will probably be a release of mathics-core 6.0.1 over the upcoming weekend to correct a missing dependency on pillow on  Python > 3.6.
